src/Adapter/Carrier/Validate/CarrierValidator.php

```diff
@@ -84,4 +88,39 @@ public function validateLogoUpload(string $filePath): void
             throw new MemoryLimitException('Cannot upload image due to memory restrictions');
         }
     }
+
+    /**
+     * @throws CoreException
+     */
+    private function validateGeneral(Carrier $carrier): void
+    {
+        $this->validateObjectModelProperty($carrier, 'name', CarrierConstraintException::class, CarrierConstraintException::INVALID_NAME);
+        $this->validateObjectModelProperty($carrier, 'grade', CarrierConstraintException::class, CarrierConstraintException::INVALID_GRADE);
+        $this->validateObjectModelProperty($carrier, 'url', CarrierConstraintException::class, CarrierConstraintException::INVALID_TRACKING_URL);
+        $this->validateObjectModelProperty($carrier, 'position', CarrierConstraintException::class, CarrierConstraintException::INVALID_POSITION);
+        $this->validateObjectModelLocalizedProperty($carrier, 'delay', CarrierConstraintException::class, CarrierConstraintException::INVALID_DELAY);
+    }
+
+    /**
+     * @throws CoreException
+     */
+    private function validateShippingGeneral(Carrier $carrier): void
+    {
+        $this->validateObjectModelProperty($carrier, 'shipping_handling', CarrierConstraintException::class, CarrierConstraintException::INVALID_SHIPPING_HANDLING);
+        $this->validateObjectModelProperty($carrier, 'is_free', CarrierConstraintException::class, CarrierConstraintException::INVALID_IS_FREE);
+        $this->validateObjectModelProperty($carrier, 'shipping_method', CarrierConstraintException::class, CarrierConstraintException::INVALID_SHIPPING_METHOD);
+        $this->validateObjectModelProperty($carrier, 'range_behavior', CarrierConstraintException::class, CarrierConstraintException::INVALID_RANGE_BEHAVIOR);
+
+        // A Carrier cannot be both shipping handling and free
+        if ($carrier->shipping_handling && $carrier->is_free) {
+            throw new CarrierConstraintException(
+                'Carrier cannot be both shipping handling and free',
+                CarrierConstraintException::INVALID_SHIPPING_HANDLING
```

Comment:
I think it's worth introducing a specific error code here `INVALID_SHIPPING_HANDLING` is used for invalid types/values But here it's different it's a domain rule that prevents setting to values true at the same time, it should have a dedicated error code so we can also handle a dedicated error message based on it